### ISSUE TITLE
NSMergeConflict during fetchAndApplyServerChangesToLocalDatabase()

### ISSUE THREAD

**Issue by tifroz** [paulw11/Seam3#48]
I am seeing some merge conflict when SMStore is sync'ing incremental changes (the sync works ok when migrating, or during the initial sync at startup). As shown in the stack capture, this happens while doing a `fetchAndApplyServerChangesToLocalDatabase()` while processing conflicts reported by the server.

The conflict error itself thrown by the `SMStoreSyncOperation.localStoreMOC` (at line 377 of `SMStoreSyncOperation`, aka `try self.localStoreMOC.saveIfHasChanges()`)

The error text goes:
```
ERROR during performSync() Error Domain=NSCocoaErrorDomain Code=133020 "(null)" UserInfo={conflictList=(
    "NSMergeConflict (0x60c000464040) for NSManagedObject (0x60000009c0c0) with objectID '0xd000000000a80002 <x-coredata://07F7B42A-6390-4758-ACFB-E9D69AAFDD9C/WTVFolder/p42>' with oldVersion = 1 and newVersion = 5 and old object snapshot = {
    index = 20;
    isSystem = 0;
    label = \"F0 16:57\";
    parent = \"0xd000000000140002 <x-coredata://07F7B42A-6390-4758-ACFB-E9D69AAFDD9C/WTVFolder/p5>\";
    profile = \"0xd000000000040004 <x-coredata://07F7B42A-6390-4758-ACFB-E9D69AAFDD9C/WTVProfile/p1>\";
    \"sm_LocalStore_EncodedValues\" = \"<null>\";
    \"sm_LocalStore_RecordID\" = \"FFFEBC01-B62F-4B4D-A377-EC5324D9941C\";
} and new cached row = {
    index = 20;
    isSystem = 0;
    label = \"F0 16:57\";
    parent = \"0xd000000000140002 <x-coredata://07F7B42A-6390-4758-ACFB-E9D69AAFDD9C/WTVFolder/p5>\";
    profile = \"0xd000000000040004 <x-coredata://07F7B42A-6390-4758-ACFB-E9D69AAFDD9C/WTVProfile/p1>\";
    \"sm_LocalStore_EncodedValues\" = \"<null>\";
    \"sm_LocalStore_RecordID\" = \"FFFEBC01-B62F-4B4D-A377-EC5324D9941C\";
}"
)}
```

I imagine my data model has a good deal to do with the error so it's for me to troubleshoot, however I could use some pointers. In particular I think I understand the role as the `localStoreMOC` as a local proxy that mirrors the data on the server, but because this is a conflict error, I imagine there are other `context` objects manipulating the data from the same local storage used by  `localStoreMOC` - and I have no idea where to find these contexts and what there roles  are?

**Comment by tifroz** [comment]
For reference, here's a screen capture of the stack trace just before the error is thrown

<img width="1435" alt="screen shot 2017-10-06 at 6 00 49 pm" src="https://user-images.githubusercontent.com/150367/31303970-1bf94a9a-aacc-11e7-8d22-bf545768a88c.png">

**Comment by paulw11** [comment]
What conflict resolution option do you have in place for the Seam3 store?  Is there a Cloud Kit conflict that is associated with this conflict?  There is only a single managed object context used by Seam3 to access the sqlite backing store.

**Comment by tifroz** [comment]
I left the default conflict resolution option in place for the Seam3 store itself, but I *think* the issue is caused by the lack of merge policy on the `localStoreMOC`. In other words, I think the initial error is thrown by `try self.fetchAndApplyServerChangesToLocalDatabase() #SMStoreSynOperation Line 113`, not by `try self.applyLocalChangesToServer() #SMStoreSynOperation Line 112`

I need to do more testing to confirm (and more reading to understand what the merge policy does exactly), but I am hoping that adding `self.localStoreMOC.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy` may address the problem.

**Comment by paulw11** [comment]
Setting a merge policy on the store will prevent the crash, since the default merge policy is to throw an error, but I am not sure why there were multiple updates to the same managed object on the backing store.

I am concerned that simply setting a trump merge policy isn't the right thing to do, although from the crash it seemed that both versions of the object were identical.

**Comment by tifroz** [comment]
I suspect the 2 objects in the error are not identical as they may seem, because the relationship attributes of `WTVFolder` (the object reported in the conflict) are not shown in the error printout.

I don't know if it's relevant (my guess is that it isn't), but `localStoreMOC`gets changed in at least a couple of places during `performSync()`:
• `applyLocalChangesToServer()` makes changes to the  (`try self.localStoreMOC.saveIfHasChanges()` at SMStoreSynOperation Line 235)
• Additional changes are made in `applyServerChangesToLocalDatabase()` (via `insertOrUpdateManagedObjects()`, `deleteManagedObjects()`)

**Comment by tifroz** [comment]
Working theory: the `insertOrUpdateManagedObjects()` method loops through `ckRecords` and derives an `NSManagedObject` for each. If he same `NSManagedObject` is referenced more than once during this loop, a conflict would occur when saving the context.

Late night here, but I will look into this first time I get a chance

**Comment by tifroz** [comment]
Scrap that, the ckRecord loop isn't the issue.

I have narrowed it down a bit: the conflict is only raised when an `NSManagedObject` modified multiples times in quick sequence by the app. So the new working theory is that the sequencing gets out of whack as changes propagate.

I am not sure this is the issue, but I am concerned that the `localStoreMOC`'s `fetch` or `save` operations don't seem to be performed on the`localStoreMOC`'s private queue via `performBlockAndWait` or `performBlock` to ensure proper sequencing ([NSManagedObjectContext Doc](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext)). Am I missing something ?

**Comment by tifroz** [comment]
No 100% sure, but it looks like the `localStoreMOC` used by `SMStoreSyncOperation` and the `backingMOC` used by the app both tap into the same persistent storage, and can act independently of each other. So if the persistent storage is written to by `localStoreMOC` while a sync  operation is in progress, we're likely to run into some conflicts.

Does that seem like a reasonable culprit candidate to you?

**Comment by paulw11** [comment]
Yes, that does sound plausible.  The symptom you are seeing definitely looks like a threading issue.  Seam3 itself only uses a single queue, but that is independent of the queue(s) used by your app.

**Comment by tifroz** [comment]
The problem went away after I made the `SMStoreSyncOperation` listen for changes in the `backingMOC` (using `NSManagedObjectContextDidSave` notification), and merge these changes whenever they occurred (as a side note this suggests that the threading issue is between the `localStoreMOC` and the `backingMOC` & leaves the application out of the picture).

Happy to submit a PR if you'd like? The alternative is to ignore the issue since it is far from deadly (things will get back in sync the next time a `sync` is triggered).

For reference, the code looks like this:

```
class SMStoreSyncOperation: Operation {
  // ...
  override func main() {
    // ...
    NotificationCenter.default.addObserver(self, selector: #selector(SMStoreSyncOperation.backingContextDidSave(notification:)), name: NSNotification.Name.NSManagedObjectContextDidSave, object: backingMOC)
  }

  @objc func backingContextDidSave(notification: Notification) {
    print("OK backingContextDidSave")
    self.localStoreMOC.performAndWait {
      self.localStoreMOC.mergeChanges(fromContextDidSave: notification)
    }
  }
}
```


I should mention that when the app makes rapid changes to a single object, the sync operations that follow the first one will still be aborted. The fix suggested above would however avoid running into a messy `NSConflictError`

**Comment by paulw11** [comment]
Please submit a PR.  

Disabling auto-sync (now that that actually works :) ) is suggested before making a large number of changes to records, but the more robust we can make the library the better.

I am wondering if instead of just skipping the sync operation is one is already in progress I should also set a flag and kick off another sync at the completion of the current one to pick up any changes that may be pending.

**Comment by tifroz** [comment]
Unless they come from a server push notification, I think all the calls `triggerSync(block: Bool, complete: Bool = false)` are made with a `false` value for the `block` parameter - which ensures subsequent calls to sync will be aborted until the current one is done.

I ignored the `block` parameter as a test (meaning I made it possible for the sync `operationQueue` to queue up more than 1 operation at a time), and everything worked out ok when I tested with 2 changes made in quick succession, each firing a `sync`.

I suppose that this `block` parameter may be a legacy piece of code that is no longer useful? It's tempting to just take it out, but not knowing the history behind it, I am not sure what I may be overlooking.

**Comment by paulw11** [comment]
The block parameter is used during processing of push notifications to ensure that the completion handler isn't called until the sync has completed.

**Comment by tifroz** [comment]
I can get rid of the block parameter, get the completion handler passed on to the syncOperation and called upon completion of the operation as part of the PR - if you don’t have any objections

**Comment by paulw11** [comment]
No, I don't believe that Seam3 should call the completion handler; that is the responsibility of the AppDelegate.

**Comment by tifroz** [comment]
Fair enough. Let's have a generic + optional completion handler with each
`triggerSync`, and let the app do what it needs to do upon completion of
the operation (in the case of a push notification, the app will use it to
call the push `completionHandler` at the appropriate time)

**Comment by paulw11** [comment]
We can add an optional completion handler, but the option for a synchronous trigger needs to remain.  As it is now you can simply call the push handling code and it will complete synchronously.  If that function becomes asynchronous then you need to implement appropriate asynchronous handling code such as `beginBackgroundTaskWithExpirationHandler` and it just makes things generally more complicated.  In most cases asynchronous execution is desirable, and this is the default, but in some cases a synchronous operation is simpler.

**Comment by tifroz** [comment]
I left all the stuff with the block parameters Vs callback out of the PR - at least for now.

I am a bit confused by your comments. The only thing the `block` parameter does, is prevent the number of `SMStoreSyncOperation`s in the `operationQueue`  to grow larger than 1 (the `SMStoreSyncOperation` still executes on a background thread asynchronously).
As an illustration, in the Example app, while the `fetchCompletionHandler` is indeed called synchronously, it is called before the operation complete (it violates the intend of the interface, because we are supposed to call the handler after we know if the operation was successful or not).

Also I am not sure what `beginBackgroundTaskWithExpirationHandler` has to do with executing tasks asynchronously? The `Background` in `beginBackgroundTaskWithExpirationHandler` refers to the application running in the background, it has nothing to do with background threads that I know of.

**Comment by paulw11** [comment]
Correct.  It is possible that a sync notification will be received while the app is suspended, causing it to begin execution in the background. As sync may take some time, it may be necessary to call `beginBackgroundTaskWithExpirationHandler`.  If the sync completes synchronously then it is easy to call the end the background task when the sync is complete.

**Comment by paulw11** [comment]
Is this issue addressed by your PR?

**Comment by tifroz** [comment]
Background processing is not addressed by the PR (currently I am leaving it up to the app to decide when to start/end a background task). This said I agree it would make sense to wrap each `triggerSync` call in a background task - happy to add it to the PR if that works for you? Let me know!

**Comment by paulw11** [comment]
Sorry, I meant this issue in general, i.e. *NSMergeConflict during fetchAndApplyServerChangesToLocalDatabase()*, not the question of background

**Comment by tifroz** [comment]
It was fixed by PR #51 ~3 weeks ago

**Comment by tifroz** [comment]
I usually leave it to the repository owner to close issues - but this one is closed as far as I am concerned.
